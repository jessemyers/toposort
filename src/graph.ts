export type Node = string;
export type Edge = [Node, Node];

export type NodeSet = Set<Node>;
export type EdgeMap = Map<Node, Set<Node>>;

export interface Graph {
    nodes: NodeSet;
    incoming: EdgeMap;
    outgoing: EdgeMap;
}

export function buildGraph(edges: Edge[]): Graph {
    const nodes = new Set(
        edges.flatMap(
            (edge) => edge,
        ),
    );

    const incoming: EdgeMap = new Map<Node, Set<Node>>();
    const outgoing: EdgeMap = new Map<Node, Set<Node>>();

    for (const node of nodes) {
        if (!incoming.has(node)) {
            incoming.set(node, new Set<Node>());
        }
        if (!outgoing.has(node)) {
            outgoing.set(node, new Set<Node>());
        }
    }

    for (const [fromNode, toNode] of edges) {
        const fromNodes = incoming.get(toNode);
        const toNodes = outgoing.get(fromNode);

        if (fromNodes) {
            fromNodes.add(fromNode);
        }
        if (toNodes) {
            toNodes.add(toNode);
        }
    }

    return {
        nodes,
        incoming,
        outgoing,
    };
}
