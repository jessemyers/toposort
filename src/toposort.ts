import { Node, Graph } from './graph';

/* Topological sort using Kahn's algorithm
 */
export function toposort({ incoming, outgoing }: Graph): Node[] {
    const sorted: Node[] = [];
    const candidates = new Set<Node>();

    // initialize candidate set with all nodes with no incoming edges
    for (const [node, edges] of incoming) {
        if (edges.size === 0) {
            candidates.add(node);
        }
    }

    while (candidates.size) {
        for (const fromNode of candidates) {
            candidates.delete(fromNode);

            // add to output list
            sorted.push(fromNode);

            // eliminate its outbound edges
            const toNodes = outgoing.get(fromNode) || [];
            for (const toNode of toNodes) {
                const edges = incoming.get(toNode);

                if (edges) {
                    edges.delete(fromNode);

                    if (edges.size === 0) {
                        candidates.add(toNode);
                    }
                }
            }
        }
    }

    return sorted;
}
