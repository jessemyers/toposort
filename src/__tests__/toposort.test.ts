import { buildGraph, Edge } from '../graph';
import { toposort } from '../toposort';

describe('toposrt', () => {
    it('sorts a graph', () => {
        const edges: Edge[] = [
            ['one', 'two'],
            ['two', 'three'],
            ['two', 'four'],
            ['three', 'four'],
        ];

        const graph = buildGraph(edges);
        const sorted = toposort(graph);

        expect(sorted).toMatchSnapshot();
    });
});
