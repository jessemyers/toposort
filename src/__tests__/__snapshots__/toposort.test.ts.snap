// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toposrt sorts a graph 1`] = `
Array [
  "one",
  "two",
  "three",
  "four",
]
`;
